source/parser/independent/intermediate-instruction.cpp

```diff
@@ -0,0 +1,259 @@
+#include "parser/independent/preprocessing-immutable-arguments.hpp"
+#include "parser/independent/section-tracker.hpp"
+#include "parser/independent/symbol-graph.hpp"
+#include "parser/independent/symbol-replacer.hpp"
+#include "parser/independent/syntax-tree-generator.hpp"
+
+using size_t = std::size_t;
+
+IntermediateInstruction::IntermediateInstruction(
+    const CodePositionInterval& positionInterval,
+    const std::vector<PositionedString>& labels,
+    const PositionedString& name,
+    const std::vector<PositionedString>& sources,
+    const std::vector<PositionedString>& targets)
+: IntermediateOperation(positionInterval, labels, name)
+, _sources(sources)
+, _targets(targets) {
+}
+
+void IntermediateInstruction::execute(
+    const ExecuteImmutableArguments& immutable,
+    CompileErrorList& errors,
+    FinalCommandVector& commandOutput,
+    MemoryAccess& memoryAccess) {
+  // For a machine instruction, it is easy to "execute" it: just insert it
+  // into the final form.
+  auto compiled = compileInstruction(immutable, errors, memoryAccess);
+  if (compiled.node()) {
+    commandOutput.push_back(compiled);
+  }
+}
+
+std::vector<std::shared_ptr<AbstractSyntaxTreeNode>>
+IntermediateInstruction::compileArgumentVector(
+    const std::vector<PositionedString>& vector,
+    const ExecuteImmutableArguments& immutable,
+    CompileErrorList& errors,
+    MemoryAccess& memoryAccess) {
+  std::vector<std::shared_ptr<AbstractSyntaxTreeNode>> output;
+
+  SymbolReplacer::DynamicReplacer labelReplacerFunction = [&](
+      const Symbol& symbol) {
+    // This might not be the biggest type, but we got to consider that the
+    // memory also works on size_t.
+    size_t byteBitSize =
+        sizeof(size_t) * immutable.architecture().getByteSize();
+    auto labelValue = conversions::convert<size_t>(
+        std::stoul(symbol.value().string()), byteBitSize);
+    auto instructionAdress =
+        conversions::convert<size_t>(_relativeAddress.offset(), byteBitSize);
+    auto relativeAdress =
+        immutable.generator().getNodeFactories().labelToImmediate(
+            labelValue, name().string(), instructionAdress);
+    return relativeAdress.toHexString(true, true);
+  };
+
+  auto replacer = SymbolReplacer(immutable.replacer(), labelReplacerFunction);
+
+  output.reserve(vector.size());
+  for (const auto& operand : vector) {
+    auto transformed =
+        immutable.generator().transformOperand(operand, replacer, errors);
+
+    // Only add argument node if creation was successful.
+    // Otherwise AbstractSyntaxTreeNode::validate() segfaults.
+    if (transformed) {
+      output.emplace_back(std::move(transformed));
+    }
+  }
+  return output;
+}
+
+FinalCommand IntermediateInstruction::compileInstruction(
+    const ExecuteImmutableArguments& immutable,
+    CompileErrorList& errors,
+    MemoryAccess& memoryAccess) {
+  // We replace all occurenced in target in source (using a copy of them).
+  auto srcCompiled =
```

Comment:
I'd prefer `sourceCompiled` and `targetCompiled` or (or `compiledSource/Target`)